File: Data Structures & Algorithms/island-perimeter/submission-1.cpp
class Solution {
public:
    int islandPerimeter(vector<vector<int>>& grid) {
        vector<vector<bool>>visited(
            grid.size(),
            vector<bool>(grid[0].size(),false)
        );
        int r = grid.size();
        int c = grid[0].size();
        for(int i = 0 ; i < r ; i++) {
            for(int j = 0 ; j < c ; j++) {
                if (grid[i][j] == 1) {
                   return dfs(grid,visited,j,i);
                }
            }
        }
    }

    int dfs(vector<vector<int>>& grid, vector<vector<bool>>& v, int x, int y) {
        int r = grid.size()-1;
        int c = grid[0].size()-1;
        if (x < 0 || x > c || y < 0 || y > r) {
            return 1;
        } else if (v[y][x]) {
            return 0;
        } else if (grid[y][x] == 0) {
            return 1;
        }else {
            v[y][x] = true;
            return dfs(grid,v,x+1,y) + dfs(grid,v,x-1,y) + 
            dfs(grid,v,x,y+1) + dfs(grid,v,x,y-1);
        }
    }
};

File: Data Structures & Algorithms/combination-target-sum/submission-1.rs
impl Solution {
    pub fn combination_sum(nums: Vec<i32>, target: i32) -> Vec<Vec<i32>> {
        let mut path = Vec::new();
        let mut res = Vec::new();
        let mut nums = nums;
        nums.sort();
        Self::dfs(0,&nums, &mut path, &mut res, 0, &target);
        res
    }

    pub fn dfs (idx: usize, nums: &[i32], path: &mut Vec<i32>, res: &mut Vec<Vec<i32>>, sum: i32, target: &i32) {
        let mut new_sum = sum;

        if new_sum == *target {
            res.push(path.clone());
            return;
        }

        for i in idx..nums.len() {
            new_sum += nums[i];
            if new_sum > *target {
                return;
            }
            path.push(nums[i]);
            Self::dfs(i,nums,path,res,new_sum,target);
            path.pop();
            new_sum -= nums[i];
        }
    }
}
